### PR TITLE
Upgrade 0932

### LINKED ISSUE #927: Upgrade to polkadot v0.9.32
Dependencies
Cumulus: https://github.com/paritytech/cumulus/tree/polkadot-v0.9.32
Moonbeam: https://github.com/PureStake/moonbeam/tree/v0.28.1
Nimbus: https://github.com/PureStake/nimbus/tree/moonbeam-polkadot-v0.9.32
ORML: https://github.com/open-web3-stack/open-runtime-module-library/tree/polkadot-v0.9.32
Polkadot: https://github.com/paritytech/polkadot/tree/release-v0.9.32
Substrate: https://github.com/paritytech/substrate/tree/polkadot-v0.9.32

Changes
Cumulus: https://github.com/paritytech/cumulus/compare/polkadot-v0.9.29...polkadot-v0.9.32
Moonbeam: https://github.com/PureStake/moonbeam/compare/v0.27.2...v0.28.1
Nimbus: https://github.com/PureStake/nimbus/compare/moonbeam-polkadot-v0.9.29...moonbeam-polkadot-v0.9.32
ORML: https://github.com/open-web3-stack/open-runtime-module-library/compare/polkadot-v0.9.29...polkadot-v0.9.32
Polkadot: https://github.com/paritytech/polkadot/compare/release-v0.9.29...release-v0.9.32
Substrate: https://github.com/paritytech/substrate/compare/polkadot-v0.9.29...polkadot-v0.9.32
Substrate-Parachain: https://github.com/substrate-developer-hub/substrate-parachain-template/compare/polkadot-v0.9.29...polkadot-v0.9.32

### PR DESCRIPTION
**Pull request by vivekvpandya**
Upgrade from 0.9.29 to 0.9.32 fixes #927 

**Notable PR**
https://github.com/paritytech/substrate/pull/11649 (has migrations)
https://github.com/paritytech/polkadot/pull/6091
https://github.com/paritytech/substrate/pull/11981
https://github.com/paritytech/substrate/pull/12109/files

**Summary of changes**
- pallet-preimage now handles preimages for pallet-democracy
- pallet-scheduler is now able to execute code stored in preimages inside pallet-preimage
- pallet-tokens offers extensives hooks:
```
            type OnDust = orml_tokens::TransferDust<Runtime, DustAccount>;
            type OnKilledTokenAccount = ();
            type OnNewTokenAccount = ();
            type OnSlash = ();
            type PostDeposit = ();
            type PostTransfer = ();
            type PreDeposit = ();
            type PreTransfer = ();
```

- (try-runtime) `pre_upgrade` now returns a vector of bytes that can be used in `post_upgrade` (deprecates `get/set_temp_storage`)
- `Origin` is now replaced by `RuntimeOrigin`
- `Call` is now replaced by `RuntimeCall`

### CODE REVIEW

**Review by sea212 — DISMISSED** [review]

File: runtime/zeitgeist/Cargo.toml [hunk 1]
```diff
@@ -115,9 +116,9 @@ zrml-swaps = { default-features = false, path = "../../zrml/swaps" }
 zrml-swaps-runtime-api = { default-features = false, path = "../../zrml/swaps/runtime-api" }
 
 [dev-dependencies]
-sp-io = { branch = "moonbeam-polkadot-v0.9.29", git = "https://github.com/zeitgeistpm/substrate" }
+sp-io = { branch = "polkadot-v0.9.32", git = "https://github.com/paritytech/substrate" }
 test-case = "2.0.2"
-xcm-emulator = { branch = "moonbeam-polkadot-v0.9.29", git = "https://github.com/zeitgeistpm/xcm-simulator" }
+xcm-emulator = { rev = "158a6bd2768c679563efa891aa17329635b2764b", git = "https://github.com/shaunxw/xcm-simulator" }
```
Comment (Chralt98) [inline runtime/zeitgeist/Cargo.toml:121]: Ok, checked. It's actually the commit with version `0.9.32` ✅

**Review by Chralt98 — DISMISSED** [review]
LGTM.

Just wanted to mention this https://github.com/zeitgeistpm/zeitgeist/issues/936 .

Both try-runtime migration tests checked ✅ .

Client connects to old nodes ✅ .

I get this test failure:

```text
failures:

---- integration_tests::xcm::tests::transfers::transfer_roc_to_relay_chain stdout ----
thread 'integration_tests::xcm::tests::transfers::transfer_roc_to_relay_chain' panicked at 'assertion failed: `(left == right)`
  left: `999988806429`,
 right: `948894198216`', runtime/battery-station/src/integration_tests/xcm/tests/transfers.rs:238:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

My command was: `cargo test --all-features`

File: runtime/common/src/lib.rs [hunk 2]
```diff
@@ -547,25 +561,35 @@ macro_rules! impl_config_traits {
             type WeightInfo = weights::orml_currencies::WeightInfo<Runtime>;
         }
 
+        pub struct CurrencyHooks<R>(sp_std::marker::PhantomData<R>);
+        impl<C: orml_tokens::Config> orml_traits::currency::MutationHooks<AccountId, CurrencyId, Balance> for CurrencyHooks<C> {
+            type OnDust = orml_tokens::TransferDust<Runtime, DustAccount>;
+            type OnKilledTokenAccount = ();
+            type OnNewTokenAccount = ();
+            type OnSlash = ();
```
[the remark this answers is not in the document]
Comment (sea212) [reply runtime/common/src/lib.rs:569]: Good catch. Fix https://github.com/zeitgeistpm/zeitgeist/pull/1005

File: node/Cargo.toml [hunk 3]
```diff
@@ -3,99 +3,99 @@ name = "zeitgeist"
...
-nimbus-consensus = { branch = "moonbeam-polkadot-v0.9.29", default-features = false, git = "https://github.com/zeitgeistpm/nimbus", optional = true }
-nimbus-primitives = { branch = "moonbeam-polkadot-v0.9.29", default-features = false, git = "https://github.com/zeitgeistpm/nimbus", optional = true }
-pallet-author-inherent = { branch = "moonbeam-polkadot-v0.9.29", default-features = false, git = "https://github.com/zeitgeistpm/nimbus", optional = true }
-pallet-parachain-staking = { tag = "v0.27.2-a", git = "https://github.com/zeitgeistpm/moonbeam", optional = true }
+moonbeam-vrf = { git = "https://github.com/zeitgeistpm/external", optional = true }
+nimbus-consensus = { default-features = false, git = "https://github.com/zeitgeistpm/external", optional = true }
+nimbus-primitives = { default-features = false, git = "https://github.com/zeitgeistpm/external", optional = true }
+pallet-author-inherent = { default-features = false, git = "https://github.com/zeitgeistpm/external", optional = true }
+pallet-parachain-staking = { default-features = false, git = "https://github.com/zeitgeistpm/external", optional = true }
```
[the remark this answers is not in the document]
Comment (sea212) [reply node/Cargo.toml:70]: The only way currently is to diff the files from the original repository and the external repository. I am aware that this is far from ideal and a request for this PR was to use forks. Given the urgency for the next release, using appropriate forks will be introduced in the next maintenance PR.

**Comment by sea212** [comment]
@Chralt98 
> failures:
> 
> [...]
> 
> My command was: `cargo test --all-features`

Interesting, seems the test script didn't consider all features. Fixed in https://github.com/zeitgeistpm/zeitgeist/pull/1006.

In regards to the failing tests: XCM instruction weights have been improved, see [this](https://github.com/paritytech/polkadot/blob/c71e872afed296d1825b15ea4b2a74750c1ba647/runtime/rococo/src/xcm_config.rs#L156-L160) and [this](https://github.com/paritytech/polkadot/blob/release-v0.9.32/runtime/rococo/src/weights/xcm/mod.rs) ([diff here](https://github.com/paritytech/polkadot/compare/release-v0.9.29-1...release-v0.9.32)). Equally true for Polkadot.

Mismatching fees fixed in https://github.com/zeitgeistpm/zeitgeist/pull/966/commits/6f82f9a43d2be7b22df02837fdcc06e03f1d156a

File: runtime/common/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Chralt98) [reply runtime/common/src/lib.rs:569]: Oh, what I initially meant was the `OnSlash` type. But anyways, cool this comment was useful for the dust too.

**Review by Chralt98 — APPROVED** [review]

File: runtime/common/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sea212) [reply runtime/common/src/lib.rs:569]: Nice coincidence, I like :)

`OnSlash` is a hook that is executed before the slashing happens: https://github.com/open-web3-stack/open-runtime-module-library/blob/4ae0372e2c624e6acc98305564b9d395f70814c0/traits/src/currency.rs#L698

Every pallet is responsible for handling slashed tokens. Perhaps the pallet does something specific with the slashed tokens (pay out other participants for example). The best practice is to add a `Slash` config trait that requires the implementation of the trait [`OnUnbalanced`](https://github.com/paritytech/substrate/blob/f4cb21c876d4ce19eb9ef09c39bebf83ba01d16d/frame/support/src/traits/tokens/imbalance/on_unbalanced.rs#L22-L55) for whenever the situation occurs that slashed tokens are not used anymore. `pallet-treasury` implements this hook, such that the slashed funds are moved into the treasury account. So one can simply configure the `Slash` config trait to the treasury instance in the runtime, e.g. `Slash = Treasury`.

**Review by Chralt98 — APPROVED** [review]

**Review by sea212 — APPROVED** [review]

**Pull request merged**